feat: Add multi-user support

raccoonasdf opened the pull request:
Adds multi-user support by 1. allowing the config to specify distinct API keys for individual cards and 2. setting the Tachi user ID and API key alongside the card ID on every `cardmng.inquire`.

This involves a nontrivial breaking change in the config format: I replace the card whitelist with card-specific API keys and an optional “default” key. You can accomplish the same behavior as the previous whitelist system, but it will be more verbose if for whatever reason you have lots of cards for one Tachi account.

I’m by no means a professional, but I’ve done a little trivial testing (shared setup, two sessions on two different cards linked to two different Tachi accounts, no reboot inbetween) and it seems to work for me.

Let me know if you’d like any changes made to the approach :)

raccoonasdf committed:
feat: Add multi-user support
Replace notion of "whitelist" with card-specific API keys
Store Card ID and Tachi user ID together in a single User struct
Acquire Tachi user on every cardmng.inquire instead of once on startup

adamaq01 commented:
Hello, thank you for putting that PR up, it is very much appreciated 🙏

I am currently on a trip without a computer on hand so I was only able to quickly look into it from my phone.
Just wanted to let you know that I will properly review this once I'm back home, hopefully I'll have time next week 👍

I mostly want to ask you to make changes to the configuration format as I've discussed with several other players and think we can have something better here.
It would be ideal to not make any changes to the way people configure things to this day and add a "profile" feature.
So you would basically still have the same base configuration file but you could add entries as such:
```TOML
[profiles.player_1]
api_key = "sample_api_key1"
cards = ["E0001...", "E0002..."]

[profiles.username2]
api_key = "sample_api_key2"
cards = ["E000..."]
```
This way, the current configuration sections would account for a "default" profile so it doesn't change anything for current users and people would be allowed to create profiles for their friends with support for multiple cards (say if one day one friend forgets their card).
One benefit of this approach is also that later on we would be able to implement profile overrides for other configuration sections I guess, or also parse profiles from totally separate files, that's also why I like it, though this is out of scope of this PR I think.

Let me know what you think, also if you don't want to make those changes that's totally okay, I can accept the PR once a proper code review is done and I'll find the time to implement this myself afterwards as your code is a really good starting point.

Cheers!

raccoonasdf commented:
Aye, that makes sense! At first wrote this just to scratch my own itch before I saw on the release notes for 0.2.3 that it was a desired feature, so I originally didn’t prioritize stability in the config format at all. Should just be surface-level changes to get it the way you suggested. Hoping I can get around to doing that by the time you can review :)

raccoonasdf committed:
restore compatibility with existing single-user configs
use seperate `[profiles.*]` tables for multi-user configuration
fallback to `cards.whitelist` and `tachi.api_key` for the "default profile"

adamaq01 commented on src/helpers.rs:
```diff
@@ -57,40 +59,50 @@ where
         err.into_inner()
     });
-
+    
```
nit: remove this

raccoonasdf replied:
whoops!

adamaq01 commented on src/helpers.rs:
```diff
@@ -57,40 +59,50 @@ where
+        .cloned()
+        // ... or fallback to single-user config, if present
+        .or_else(|| {
+            let cards_config = CONFIGURATION.cards.as_ref()?;
```
Let's change this `?` with an `unwrap_or_default` wdyt?

raccoonasdf replied:
since omitting `[cards]` from the config is not valid in current release, i took the liberty of making this omission a signal that the user doesn't care about the single-user behavior at all, just `[profiles]`. because `[cards]` with an empty whitelist means "wildcard", i wanted a state that means we don't want any card to be considered.

of course, there's another `?` on the very next line, so i can still get the behavior i want regardless.

if you do still want "omitted `[cards]`" to behave the same as "default `[cards]`", may i suggest instead that i change `cards` back into a non-Optional `CardsConfiguration` while keeping the `#[serde(default)]`?

adamaq01 commented on src/mikado.rs:
```diff
@@ -311,11 +308,50 @@ pub unsafe fn property_destroy_hook(property: *mut ()) -> i32 {
+                })
+            });
+
+            *guard = user;
```
Assign directly to guard
You can rename guard to user if you want and prefer it that way

raccoonasdf replied:
agreed.

adamaq01 reviewed (approved):
Thanks a lot, I asked for some small changes but it is mainly minor stuff.

I am not super happy with how some things are done but it is mostly my fault due to how I made the static data stuff initially so it would be annoying to ask you to change these.
I'll try to cleanup whenever I have some spare time/motivation for the project, but this is a great contribution/addition and the code itself is well written 🙏 

I'll merge once the comments are addressed and publish a new release 👍

raccoonasdf committed:
cleanup after review

adamaq01 commented:
Hey, it took me some time to find the motivation to come back to this. Sorry for the wait.
I've made some slight changes before merging as I think some types were better off living elsewhere.

Thanks for your contribution!
